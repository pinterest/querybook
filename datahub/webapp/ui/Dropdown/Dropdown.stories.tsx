import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import centered from '@storybook/addon-centered/react';

import { Dropdown } from './Dropdown';
import { ListMenu } from 'ui/Menu/ListMenu';
import { Button } from 'ui/Button/Button';

export default {
    title: 'Stateful/Dropdown',
    decorators: [centered],
};
export const _Dropdown = (args) => (
    <>
        <Dropdown {...args}>
            <ListMenu
                items={[{ name: 'Dropdown Item' }, { name: 'with ListMenu' }]}
            />
        </Dropdown>
        <div className="mb12" />
        <Dropdown
            {...args}
            customButtonRenderer={() => (
                <Button icon={'heart'} title="Dropdown" />
            )}
        >
            <ListMenu
                items={[
                    { name: 'Soft Dropdown' },
                    { name: 'with Custom Button' },
                    {
                        name: 'Nested Item',
                        items: [
                            { name: 'Nested Option 1' },
                            { name: 'Nested Option 2' },
                        ],
                    },
                ]}
                soft
            />
        </Dropdown>
    </>
);

_Dropdown.args = {
    menuIcon: 'menu',
    hoverable: false,
    isRight: false,
    isUp: false,
};
