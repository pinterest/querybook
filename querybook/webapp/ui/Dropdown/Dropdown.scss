.Dropdown {
    position: relative;
    display: flex;
    flex-direction: column;
    justify-content: center;

    .Dropdown-trigger {
        cursor: pointer;
        display: flex;
    }

    .Dropdown-menu {
        left: 0;
        min-width: 80px;
        position: absolute;
        top: 100%;
        z-index: 20;
        .Menu {
            border: var(--border);
            border-radius: var(--border-radius);
        }
    }

    &.is-right {
        .Dropdown-menu {
            left: auto;
            right: 0;
        }
    }
    &.is-up {
        .Dropdown-menu {
            bottom: 100%;
            padding-bottom: 4px;
            padding-top: initial;
            top: auto;
        }
    }

    &.Dropdown-nested-menu {
        .Dropdown-menu {
            top: -1px;
            // this right now doesn't work
            margin-left: 100%;
        }
    }
    &.Dropdown-nested-menu.nested-right {
        .Dropdown-menu {
            margin-right: 100%;
        }
    }
}
