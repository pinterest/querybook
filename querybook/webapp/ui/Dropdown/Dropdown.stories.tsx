import centered from '@storybook/addon-centered/react';
import React from 'react';

import { Button } from 'ui/Button/Button';
import { ListMenu } from 'ui/Menu/ListMenu';

import { Dropdown } from './Dropdown';

export default {
    title: 'Stateful/Dropdown',
    decorators: [centered],
};
export const _Dropdown = (args) => (
    <>
        <Dropdown {...args}>
            <ListMenu
                items={[{ name: 'Dropdown Item' }, { name: 'with ListMenu' }]}
            />
        </Dropdown>
        <div className="mb12" />
        <Dropdown
            {...args}
            customButtonRenderer={() => (
                <Button icon={'ChevronDown'} title="Dropdown" />
            )}
        >
            <ListMenu
                items={[
                    { name: 'Soft Dropdown' },
                    { name: 'with Custom Button' },
                    {
                        name: 'Nested Item',
                        items: [
                            { name: 'Nested Option 1' },
                            { name: 'Nested Option 2' },
                        ],
                    },
                ]}
                soft
            />
        </Dropdown>
    </>
);

_Dropdown.args = {
    menuIcon: 'menu',
    hoverable: false,
    layout: ['bottom', 'right'],
};
