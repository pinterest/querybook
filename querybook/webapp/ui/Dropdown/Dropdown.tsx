import clsx from 'clsx';
import React from 'react';

import { IconButton } from 'ui/Button/IconButton';
import { Popover } from 'ui/Popover/Popover';

import './Dropdown.scss';

interface IProps {
    menuIcon?: string;
    className?: string;
    customButtonRenderer?: () => React.ReactNode;

    hoverable?: boolean;
    isRight?: boolean;
    isUp?: boolean;

    usePortal?: boolean;
}

export const Dropdown: React.FunctionComponent<IProps> = ({
    menuIcon = 'menu',
    className,
    customButtonRenderer,

    hoverable = true,
    isRight,
    isUp,
    usePortal,

    children,
}) => {
    const selfRef = React.useRef<HTMLDivElement>(null);
    const popoverRef = React.useRef<HTMLDivElement>(null);

    const [active, setActive] = React.useState(false);
    // Hover Based Dropdown control
    const handleMouseEnter = React.useCallback(() => {
        if (hoverable) {
            setActive(true);
        }
    }, [hoverable]);
    const handleMouseExit = React.useCallback(() => {
        if (hoverable) {
            setActive(false);
        }
    }, [hoverable]);

    // Click based Dropdown Control
    const handleClick = React.useCallback((event) => {
        if (!hoverable) {
            event.stopPropagation();
            event.preventDefault();
            setActive(true);
        }
    }, []);
    const onDocumentClick = React.useCallback(
        (event: MouseEvent) => {
            if (
                !(
                    event.composedPath().includes(selfRef.current) ||
                    (popoverRef.current &&
                        event.composedPath().includes(popoverRef.current))
                )
            ) {
                setActive(false);
            }
        },
        [selfRef.current]
    );

    React.useEffect(() => {
        if (!hoverable && active) {
            document.addEventListener('mousedown', onDocumentClick);
        }
        return () => {
            if (!hoverable && active) {
                document.removeEventListener('mousedown', onDocumentClick);
            }
        };
    }, [hoverable, active, onDocumentClick]);

    const buttonDOM = customButtonRenderer ? (
        customButtonRenderer()
    ) : (
        <IconButton icon={menuIcon} />
    );

    const customFormatClass = customButtonRenderer ? ' custom-format ' : '';

    const combinedClassName = clsx({
        Dropdown: true,
        [className]: className,
        [customFormatClass]: true,
        'is-right': isRight,
        'is-up': isUp,
    });

    let dropdownContent =
        active && children ? (
            <div className="Dropdown-menu" role="menu">
                {children}
            </div>
        ) : null;
    if (usePortal && dropdownContent) {
        dropdownContent = (
            <Popover
                ref={popoverRef}
                onHide={() => setActive(false)}
                anchor={selfRef.current}
                layout={[isUp ? 'top' : 'bottom', isRight ? 'right' : 'left']}
                hideArrow
                skipAnimation
            >
                {dropdownContent}
            </Popover>
        );
    }

    return (
        <div
            className={combinedClassName}
            onMouseEnter={handleMouseEnter}
            onMouseLeave={handleMouseExit}
            onClick={handleClick}
            ref={selfRef}
        >
            <div className="Dropdown-trigger">{buttonDOM}</div>
            {dropdownContent}
        </div>
    );
};
