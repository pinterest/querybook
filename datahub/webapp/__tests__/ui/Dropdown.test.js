import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';

import { Dropdown } from '../../ui/Dropdown/Dropdown';

it('renders without crashing', () => {
    shallow(<Dropdown />);
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(<Dropdown />);
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});

describe('matches test renderer snapshot', () => {
    it('serializes the styles', () => {
        const output = renderer.create(<Dropdown />);
        expect(output).toMatchSnapshot();
    });
});
