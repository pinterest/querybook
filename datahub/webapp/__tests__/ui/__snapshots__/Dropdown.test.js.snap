// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot 1`] = `
<div
  className="Dropdown "
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="Dropdown-trigger"
  >
    <ForwardRef
      icon="menu"
    />
  </div>
</div>
`;

exports[`matches test renderer snapshot serializes the styles 1`] = `
<div
  className="Dropdown "
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="Dropdown-trigger"
  >
    <span
      className="IconButton"
    >
      <span
        className=" Icon "
        dangerouslySetInnerHTML={
          Object {
            "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" class=\\"feather feather-menu\\"><line x1=\\"3\\" y1=\\"12\\" x2=\\"21\\" y2=\\"12\\"></line><line x1=\\"3\\" y1=\\"6\\" x2=\\"21\\" y2=\\"6\\"></line><line x1=\\"3\\" y1=\\"18\\" x2=\\"21\\" y2=\\"18\\"></line></svg>",
          }
        }
      />
       
    </span>
  </div>
</div>
`;
